029-across-step/proposal.md

```diff
@@ -0,0 +1,277 @@
+# `across` step
+
+This proposal introduces a mechanism by which a build plan can be executed
+multiple times with different `((vars))`, in service of build matxies, pipeline
+matrixes, and - building on [`((var))` sources][var-sources-rfc] - dynamic
+variations of both.
+
+
+## Motivation
+
+* Support dynamic multi-branch workflows:
+  [concourse/concourse#1172][multi-branch-issue]
+* Support build matrixes and pipeline matrixes.
+* Replace a common, and very flawed, usage pattern of `version: every`, which
+  has proven to be very complicated to support.
+
+
+## Proposal
+
+The `across` step modifier is given a list containing vars and associated
+values to set when executing the step. The step will be executed across all
+combinations of var values.
+
+### Static var values
+
+The set of values can be static, as below:
+
+```yaml
+task: unit
+vars: {go_version: ((.:go_version))}
```

Comment:
This is more of a question about the var steps RFC, but since that's merged, I'll put my question here.

The ["var steps" RFC](https://github.com/concourse/rfcs/blob/master/027-var-steps/proposal.md#proposal) states:

> Each build will have a single local scope which steps can use for setting and getting var values

Does this mean that the var fetched from a `load_var` (or `get_var`) step nested under e.g. a `do` block will be accessible outside of that `do` block?

```yaml
plan:
- do:
  - load_var: my-var
    file: some-file.yml
- task:
  vars:
    my-var: ((.:my-var)) # Is this allowed? My guess is yes, but just checking
```

If so, how does this work with the `across` step? Presumably, each "iteration" of the base step would need its own local scope (at least for the vars that get iterated over):

```yaml
plan:
- across:
  - var: version
    values: [v1, v2, v3]
    max_in_flight: all
  do:
  - task: some-task
    vars:
      version: ((.:version)) # ((.:version)) has to take on 3 different values, which suggests multiple scopes
```

Additionally, are the var steps allowed to be used from inside the across step? (borrowing the modified `get_var` syntax from https://github.com/concourse/rfcs/pull/29#discussion_r452312783)

```yaml
plan:
- across:
  - var: pr
    source: booklit-prs
    max_in_flight: all
  do:
  - get_var: pr-details
    path: ((.:pr.number))
    source: booklit-prs
  - task: some-task
    vars:
      assignees: ((.:pr-details.assignees))
- task: another-task
  vars:
    assignees: ((.:pr-details.assignees)) # presumably, pr-details would not be in scope here
```

Reply:
Haha, yep, good catch. I think we'll need some sort of scoping mechanism for the `across` step to have any hope of functioning in a sane way.

I think the `across` step should perhaps run its sub-steps with their own scope that inherits from the outer one and has their vars set there. That way they can still reference outer vars, and their own vars, and any vars they set within the step (like your last example) would only be reachable from within their scope - i.e. they won't "leak" to the outside.

I'll add a section for this! The var steps RFC doesn't really cover it because there isn't any notion of sub-scopes at the moment - this would be the first instance of it.